cmd/in-place-restart-agent/main.go

```diff
@@ -0,0 +1,351 @@
+	IsBarrierActive          bool
+	Exit                     func(int)                           // Required for testing
+	StartWorker              func(context.Context, string) error // Required for testing
+}
+
+// NewInPlaceRestartAgent creates a new InPlaceRestartAgent
+func NewInPlaceRestartAgent(client client.Client, namespace string, podName string, workerCommand string, inPlaceRestartExitCode int) *InPlaceRestartAgent {
+	return &InPlaceRestartAgent{
+		Client:                   client,
+		Namespace:                namespace,
+		PodName:                  podName,
+		InPlaceRestartExitCode:   inPlaceRestartExitCode,
+		WorkerCommand:            workerCommand,
+		PodInPlaceRestartAttempt: nil,
+		IsBarrierActive:          true,
+		Exit:                     os.Exit,
+		StartWorker: func(ctx context.Context, command string) error {
+			shell := os.Getenv("SHELL")
+			if shell == "" {
+				shell = "/bin/sh"
+			}
+			cmd := exec.CommandContext(ctx, shell, "-c", command)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			return cmd.Run()
+		},
+	}
+}
+
+// SetupWithManager sets up the in-place restart agent with the manager
+// Only trigger reconcile when the associated JobSet is affected
+func (r *InPlaceRestartAgent) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&jobset.JobSet{}).
+		Complete(r)
+}
+
+// Reconcile handles the in-place restart logic at the Pod level
+func (r *InPlaceRestartAgent) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("namespace", r.Namespace, "podName", r.PodName, "inPlaceRestartExitCode", r.InPlaceRestartExitCode, "workerCommand", r.WorkerCommand, "podInPlaceRestartAttempt", r.PodInPlaceRestartAttempt, "isBarrierActive", r.IsBarrierActive)
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.Info("reconciling")
+
+	// Get associated JobSet
+	var js jobset.JobSet
+	if err := r.Get(ctx, req.NamespacedName, &js); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	currentInPlaceRestartAttempt := js.Status.CurrentInPlaceRestartAttempt
+	previousInPlaceRestartAttempt := js.Status.PreviousInPlaceRestartAttempt
+	log = log.WithValues("currentInPlaceRestartAttempt", currentInPlaceRestartAttempt, "previousInPlaceRestartAttempt", previousInPlaceRestartAttempt)
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Handle start up
+	// The Pod in-place restart attempt is set only once per agent execution, instead of possibly setting it multiple times in different reconciliations
+	// This could technically be done in the main function by getting the associated JobSet and patching the Pod, but it is done in the reconcile function to reuse the manager client and its watch
+	// This is done to decrease the impact on the API server as much as possible
+	// The startup is indicated by the Pod in-place restart attempt being not being set yet
+	if r.PodInPlaceRestartAttempt == nil {
```

Comment:
Are we going to keep one Reconcile for agent? If yes, you can store this as a global variable, to not place it to the InPlaceRestartAgent struct, but it is up to you.